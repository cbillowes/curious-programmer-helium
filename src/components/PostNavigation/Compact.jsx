import React, { Component } from "react"
import Link from "gatsby-link"
import "./Compact.scss"

class CompactPostNavigation extends Component {
  render() {
    const { previous, next } = this.props

    return (
      <div className="post-navigation compact">
        <Link className="previous" to={previous.slug}><span className="title">{previous.title}</span></Link>
        <Link className="next" to={next.slug}><span className="title">{next.title}</span></Link>
      </div>
    )
  }
}

export default CompactPostNavigation
