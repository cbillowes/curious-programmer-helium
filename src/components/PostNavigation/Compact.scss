@import "../../layout/variables.scss";

.post-navigation.compact {
    font-size: 0.7em;
    font-family: $font-family-sans-serif;
    margin-top: 7em;
    height: 2.5em;

    @include small-media {
        line-height: $font-size-medium;
        height: 3.5em;
    }

    a {
        color: $foreground;

        &:hover {
            color: $secondary;
        }
    }

    .previous {
        float: left;
        margin-right: 1em;

        @include small-media {
            margin-top: 1em;
        }

        .title {
            white-space: nowrap;
            overflow: hidden;
            text-overflow: ellipsis;
            max-width: 450px;
            display: inline-block;

            &:before {
                content: "ðŸ¡ ";
                margin-right: 0.5rem;
            }

            @include medium-media {
                mex-width: 250px;
            }
        }
    }

    .next {
        float: right;
        margin-left: 1em;
        text-align: right;

        @include small-media {
            margin-top: 1em;
        }

        .title {
            white-space: nowrap;
            overflow: hidden;
            text-overflow: ellipsis;
            max-width: 450px;
            display: inline-block;

            @include medium-media {
                max-width: 250px;
            }
        }

        &:after {
            content: " ðŸ¡¢";
            float: right;
            margin-left: 0.5rem;
        }
    }

    &::after {
        content: "";
        clear: both;
        display: block;
    }
}

@include small-media {
    .post-navigation.compact {
        .next,
        .previous {
            display: block;
            float: none;
            text-align: right;
        }
    }
}
