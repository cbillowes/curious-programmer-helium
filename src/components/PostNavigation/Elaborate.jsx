import React, { Component } from "react"
import { navigate } from "gatsby"
import Metadata from "../Metadata/Post/Metadata"
import "./Elaborate.scss"

const Item = ({ post }) => (
  <aside className="post" style={{ backgroundImage: (post.cover) ? `url(${post.cover})` : "" }}>
    <div class="overlay"></div>
    <div className="container">
      <h1>{post.title}</h1>
      <p dangerouslySetInnerHTML={{ __html: post.excerpt }} />
      <Metadata tags={post.tags} date={post.date} timeToRead={post.timeToRead} />
    </div>
  </aside>
);

class ElaboratePostNavigation extends Component {
  render() {
    const { previous, next } = this.props

    return (
      <div className="post-navigation elaborate">
        <div className="nav previous" onClick={() => navigate(previous.slug)}>
          <Item post={previous} />
        </div>
        <div className="nav next" onClick={() => navigate(next.slug)}>
          <Item post={next} />
        </div>
      </div>
    )
  }
}

export default ElaboratePostNavigation
