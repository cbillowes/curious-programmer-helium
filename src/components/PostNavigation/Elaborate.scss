@import "../../layout/layout.scss";

$background-static-state: $background;

@mixin post() {
  background-size: cover;
  background-position: center;
  background-color: $background;
}

.post-navigation.elaborate {
  display: flex;

  .nav {
    cursor: pointer; 
    
    &.previous {
      @include post();
      flex: 1;

      .post-metadata p,
      .tags {
        text-align: left;
      }
    }

    &.next {
      @include post();
      flex: 1;

      .post-metadata p,
      .tags {
        text-align: right;
      }
    }

    &::after {
      content: "";
      clear: both;
      display: block;
    }

    .post {
      position: relative;

      &:hover {
        &::after {
          background-color: $background-de-emphasis;
        }
      }

      &::after {
        content: "";
        display: block;
        background-color: $background-static-state;
        opacity: 0.7;
        position: absolute;
        top: 0;
        left: 0;
        right: 0;
        bottom: 0;
      }

      .container {
        z-index: 1;
        position: absolute;
      }

      color: $foreground;
      background-color: $background-de-emphasis;
      opacity: 1;
      padding: 2em 3em;

      display:block;
      width: 100%;
      max-width: 100%;
      max-height: 500px;
      height: 100vh;
      overflow-y: auto;

      h1 {
        font-size: 1.5em;
      }
  
      a {
        color: $foreground;
      }
  
      .tag {
        background-color: $foreground;
        color: $background-de-emphasis;
        font-size: 0.7em;
        padding: 0em 1em;

        &:hover {
          background-color: $secondary;
        }
      }
    }
  }

  @media screen and (min-width: 841px) and (max-width: 1300px) {
    & {
      display :block;

      .nav.next {
        text-align: right;
      }
    }
  }

  @media screen and (max-width: 840px) {
    & {
      display: block;

      .nav {
        overflow: hidden;
        max-height: 350px;

        .post {
          padding: 0px;
          
          h1 {
            font-size: 1.25em;
          }

          h1 + p {
            display: none;
          }

          .post-metadata p,
          .post-metadata .tags {
            text-align: center;
          }
        }
      }
    }
  }
}