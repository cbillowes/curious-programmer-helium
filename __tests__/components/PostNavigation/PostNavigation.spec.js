import React from "react"
import renderer from "react-test-renderer"
import { mount } from "enzyme"
import Compact from "../../../src/components/PostNavigation/Compact"
import Elaborate from "../../../src/components/PostNavigation/Elaborate"

describe("post navigation components", () => {
  const previous = { slug: "/blog/world", title: "World", cover: "world.jpg", excerpt: "<p>World</p>", tags: ["world"] }
  const next = { slug: "/blog/galaxy", title: "galaxy", cover: "galaxy.jpg", excerpt: "<p>Galaxy</p>", tags: ["galaxy"] }

  describe("an elaborate component", () => {
    const element = <Elaborate previous={previous} next={next} />

    it("should render", () => {
      const tree = renderer.create(element).toJSON()
      expect(tree).toMatchSnapshot()
    })

    it("should contain to nav post items", () => {
      const component = mount(element)
      expect(component.find(".nav.previous .post").length).toEqual(1)
      expect(component.find(".nav.next .post").length).toEqual(1)
    })

    it("should have posts with backgrounds", () => {
      const component = mount(element)
      expect(component.find(".nav.previous .post").html()).toContain(`class="post" style="background-image: url(${previous.cover});">`)
      expect(component.find(".nav.next .post").html()).toContain(`class="post" style="background-image: url(${next.cover});">`)
    })

    it("should have posts with titles", () => {
      const component = mount(element)
      expect(component.find(".nav.previous .post").html()).toContain(`<h1>${previous.title}</h1>`)
      expect(component.find(".nav.next .post").html()).toContain(`<h1>${next.title}</h1>`)
    })

    it("should have posts with excerpts", () => {
      const component = mount(element)
      expect(component.find(".nav.previous .post").html()).toContain(`<p>${previous.excerpt}</p>`)
      expect(component.find(".nav.next .post").html()).toContain(`<p>${next.excerpt}</p>`)
    })

    it("should have posts with metadata", () => {
      const component = mount(element)
      expect(component.find(".nav.previous .post").html()).toContain(`<a class="tag" href="/tags/${previous.tags[0]}">${previous.tags[0]}</a>`)
      expect(component.find(".nav.next .post").html()).toContain(`<a class="tag" href="/tags/${next.tags[0]}">${next.tags[0]}</a>`)
    })
  })

  describe("a compact component", () => {
    const element = <Compact previous={previous} next={next} />

    it("should render", () => {
      const tree = renderer.create(element).toJSON()
      expect(tree).toMatchSnapshot()
    })

    it("should contain to nav items", () => {
      const component = mount(element)
      expect(component.html()).toContain(`<div class="post-navigation compact">`)
      expect(component.html()).toContain(`<a class="previous" href="${previous.slug}">${previous.title}</a>`)
      expect(component.html()).toContain(`<a class="next" href="${next.slug}">${next.title}</a>`)
    })
  })
})