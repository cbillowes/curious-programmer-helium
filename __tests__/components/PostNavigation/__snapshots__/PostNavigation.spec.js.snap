// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`post navigation components a compact component should render 1`] = `
<div
  className="post-navigation compact"
>
  <a
    className="previous"
    href="/blog/world"
    onClick={[Function]}
    onMouseEnter={[Function]}
  >
    World
  </a>
  <a
    className="next"
    href="/blog/galaxy"
    onClick={[Function]}
    onMouseEnter={[Function]}
  >
    galaxy
  </a>
</div>
`;

exports[`post navigation components an elaborate component should render 1`] = `
<div
  className="post-navigation elaborate"
>
  <div
    className="nav previous"
    onClick={[Function]}
  >
    <aside
      className="post"
      style={
        Object {
          "backgroundImage": "url(world.jpg)",
        }
      }
    >
      <div
        className="container"
      >
        <h1>
          World
        </h1>
        <p
          dangerouslySetInnerHTML={
            Object {
              "__html": "<p>World</p>",
            }
          }
        />
        <div
          className="post-metadata"
        >
          <p>
            posted 
             
            <time
              dateTime={"2019-01-05T08:01:38.270Z"}
            >
              a few seconds ago
            </time>
             
             
            on 
             
            <time
              dateTime={"2019-01-05T08:01:38.271Z"}
            >
              Sat, 5 January 2019
            </time>
             
             
            (Est. 
             minute read)
          </p>
          <ul
            className="tags"
          >
            <li>
              <a
                className="tag"
                href="/tags/world"
                onClick={[Function]}
                onMouseEnter={[Function]}
              >
                world
              </a>
            </li>
          </ul>
        </div>
      </div>
    </aside>
  </div>
  <div
    className="nav next"
    onClick={[Function]}
  >
    <aside
      className="post"
      style={
        Object {
          "backgroundImage": "url(galaxy.jpg)",
        }
      }
    >
      <div
        className="container"
      >
        <h1>
          galaxy
        </h1>
        <p
          dangerouslySetInnerHTML={
            Object {
              "__html": "<p>Galaxy</p>",
            }
          }
        />
        <div
          className="post-metadata"
        >
          <p>
            posted 
             
            <time
              dateTime={"2019-01-05T08:01:38.277Z"}
            >
              a few seconds ago
            </time>
             
             
            on 
             
            <time
              dateTime={"2019-01-05T08:01:38.277Z"}
            >
              Sat, 5 January 2019
            </time>
             
             
            (Est. 
             minute read)
          </p>
          <ul
            className="tags"
          >
            <li>
              <a
                className="tag"
                href="/tags/galaxy"
                onClick={[Function]}
                onMouseEnter={[Function]}
              >
                galaxy
              </a>
            </li>
          </ul>
        </div>
      </div>
    </aside>
  </div>
</div>
`;
